Update UniqueValidator.php

PowerGamer1 opened the pull request:
| Q | A |
| --- | --- |
| Is bugfix? | yes |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #11322 |
- [x] Code
- [x] Reviewed
- [x] Changelog
- [x] Tests (@SilverFire)

Fix for https://github.com/yiisoft/yii2/issues/11322
Needs many language translations for new error message: 'Combination [{valueCombo}] of [{attributeCombo}] has already been taken.'

PowerGamer1 committed:
Update UniqueValidator.php

Fix for https://github.com/yiisoft/yii2/issues/11322

PowerGamer1 committed:
Update UniqueValidator.php

Fixed error message `$comboNotUnique` to include model attribute labels instead of attribute names.

Faryshta commented:
Can you add that line to the translations?

Faryshta commented on framework/validators/UniqueValidator.php:
```diff
@@ -69,6 +78,9 @@ public function init()
             $this->message = Yii::t('yii', '{attribute} "{value}" has already been taken.');
         }
+	    if ($this->comboNotUnique === null) {
+		    $this->comboNotUnique = Yii::t('yii', 'Combination [{valueCombo}] of [{attributeCombo}] has already been taken.');
```
Can you add this to translations messages?

PowerGamer1 replied:
@Faryshta
> Can you add this to translations messages?

I can't translate this from English into other languages, so I cannot add it.

Faryshta replied:
just add it to the english translation, then the translators (including myself) will use that to translate it to other languages.

PowerGamer1 replied:
"messages" folder of the framework does not have a subfolder for English language. I think it is sufficient to specify English strings directly in the code using `Yii::t()` and then "messages/XX/yii.php" files for each language XX are generated automatically using `yii message` command line command.

SilverFire commented on framework/validators/UniqueValidator.php:
```diff
@@ -58,6 +58,15 @@ class UniqueValidator extends Validator
+	/**
+     * @var string the user-defined error message used when `targetAttribute` is an array. It may contain the following placeholders which
+     * will be replaced accordingly by the validator:
+     *
```
This line is redundant

PowerGamer1 replied:
Makes sense, but just so you know I copy-pasted this from `public $message` in `\yii\validators\Validator` so you may want to fix it there (and probably many other validator classes) too.

SilverFire commented on framework/validators/UniqueValidator.php:
```diff
@@ -58,6 +58,15 @@ class UniqueValidator extends Validator
      */
     public $filter;
+	/**
+     * @var string the user-defined error message used when `targetAttribute` is an array. It may contain the following placeholders which
```
<code>\`targetAttribute\`</code> -> `[[$targetAttribute]]`

SilverFire commented on framework/validators/UniqueValidator.php:
```diff
@@ -133,7 +145,24 @@ public function validateAttribute($model, $attribute)
 
         if ($exists) {
-            $this->addError($model, $attribute, $this->message);
+            if (is_array($targetAttribute)) {
```
The code bellow should be moved into a separate protected method to make its modification easier. And current method is already too long

SilverFire commented on framework/validators/UniqueValidator.php:
```diff
@@ -133,7 +145,24 @@ public function validateAttribute($model, $attribute)
         if ($exists) {
-            $this->addError($model, $attribute, $this->message);
+            if (is_array($targetAttribute)) {
+                $attributeCombo = [];
```
The code bellow should be moved into a separate protected method to make its modification easier. And current method is already too long

PowerGamer1 replied:
I can agree if you can come up with a good name for this separate method (`addError` cannot be used since its already defined in base class) and IMO this method should be made private so that potential refactorings of `UniqueValidator` class can have more freedom in splitting/merging class methods, etc., without worrying of breaking backward compatibility if someone used this `protected` method in their code.

SilverFire replied:
How about `addCompositeError`? Then it makes sense to rename `comboNotUnique` -> `compositeMessage`

PowerGamer1 replied:
Hm, can't get rid of word "combo" everywhere in the same way: there are also message placeholders {attributeCombo} and {valueCombo}. Saying {attributeComposition} and {valueComposition} sounds wrong. Besides, `comboNotUnique` is consistent with other validators names like `tooBig`, `wrongMimeType`, `ipv4NotAllowed`, etc. And `compositeMessage` would break the consistency. So I will name the method `addComboNotUniqueError`.

SilverFire commented on framework/validators/UniqueValidator.php:
```diff
@@ -133,7 +145,24 @@ public function validateAttribute($model, $attribute)
+            if (is_array($targetAttribute)) {
+                $attributeCombo = [];
+                $valueCombo = [];
+                foreach ($targetAttribute as $k => $v) {
```
According to [core code style](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md#42-properties) we prefer to avoid short variable names such as `$k`, `$v`. Could you rename it, please?

PowerGamer1 replied:
No problem, but it would be totally inconsistent to have here full names and just a few lines above in the same method in the similar loop I copy pasted to have short names breaking your core code style.

SilverFire commented on framework/validators/UniqueValidator.php:
```diff
@@ -133,7 +145,24 @@ public function validateAttribute($model, $attribute)
+                        $valueCombo[] = $model->$k;
+                    }
+                }
+                array_walk($attributeCombo, function (&$attribute) use($model) { $attribute = $model->getAttributeLabel($attribute); });
```
I'd prefer to put this logic in a previous loop

SilverFire commented on framework/validators/UniqueValidator.php:
```diff
@@ -133,7 +145,24 @@ public function validateAttribute($model, $attribute)
+                    }
+                }
+                array_walk($attributeCombo, function (&$attribute) use($model) { $attribute = $model->getAttributeLabel($attribute); });
+                array_walk($valueCombo, function (&$value) { $value = '"' . $value . '"'; });
```
same here

PowerGamer1 replied:
The reason I coded it outside a loop was to prevent duplication of quoting char (") for value but for performance reasons I can agree that it may be better to avoid `array_walk` and closure and put it inside a loop.

SilverFire commented:
We also need a CHANGELOG line and a unit test

PowerGamer1 committed:
Update UniqueValidator.php

Fixed most lines according to @SilverFire notes.

klimov-paul commented:
Probably, same changes should be applied to `ExistValidator`.

PowerGamer1 commented:
@SilverFire
> We also need a CHANGELOG line and a unit test

Changelog can read along the lines "Fixed incorrect error message in UniqueValidator class when multiple model attributes must constitute a unique combination" but I am not sure were to put it - feel free to rephrase and put it where needed.

As for the unit tests - no new tests should be required since no new ways of using `UniqueValidator` class were added and only its error output in some existing use cases was fixed (In other words existing unit tests need to be changed to expect a different error message added on a model from `UniqueValidator`). As I never dealt with unit testing before I leave it up to you to deal with them in whatever way you deem necessary.

Also, I fixed most lines you commented on in my pull request (you may still want to read my comments on the fixed lines though). If it makes things faster/easier feel free to copy over all the code from my pull request and make your own with whatever changes you want.

SilverFire commented:
I'd suggest the following wording: <code>Bug #11322: Fixed incorrect error message in \`yii\validators\UniqueValidator\` for composite target attribute (PowerGamer1, silverfire)</code>. It must be added to https://github.com/yiisoft/yii2/blob/master/framework/CHANGELOG.md


> As I never dealt with unit testing before I leave it up to you to deal with them in whatever way you deem necessary.

Tests are required because we are introducing new logic. I'll add them before merge.

PowerGamer1 committed:
Update UniqueValidator.php

Moved block of code from `validateAttribute` method into a separate `addComboNotUniqueError` method.

PowerGamer1 committed:
Update CHANGELOG.md

Added description for bug https://github.com/yiisoft/yii2/issues/11322.

PowerGamer1 committed:
Update UniqueValidator.php

PowerGamer1 committed:
Merge pull request #1 from PowerGamer1/patch-4

Patch 4

cebe commented:
> Probably, same changes should be applied to ExistValidator.

this does not apply to ExistValidator as that one simply talks about an attribute being invalid if validation fails.

cebe commented:
Merged with small adjustments. Thank you!

rob006 commented:
This breaks BC - if someone has custom message specified in `message` property, it will be ignored and default message for `comboNotUnique` will be used.

chathurasudarsha commented:
this will not be used since 2.1.  need a promised fix here. something like message =>['attribute1'=>'message1', 'attribute2'=>'message2'];

cebe commented:
@chathurasudarsha it is fixed in #12305
